modules/platform/nuxeo-easyshare-core/src/main/java/org/nuxeo/easyshare/EasyShare.java

```diff
@@ -91,59 +98,159 @@ public Object run(CoreSession session, IdRef docRef) throws NuxeoException {
+                                                                                                   // to
+                                                                                                   // non-collection
+                                    .arg("currentPageIndex", paginable.getCurrentPageIndex())
+                                    .arg("numberOfPages", paginable.getNumberOfPages())
+                                    .arg("docShare", docShare)
+                                    .arg("docList", paginable)
+                                    .arg("previousPageAvailable", paginable.isPreviousPageAvailable())
+                                    .arg("nextPageAvailable", paginable.isNextPageAvailable())
+                                    .arg("currentPageStatus",
+                                            paginable.getProvider().getCurrentPageStatus())
+                                    .arg("accessSecured", StringUtils.isNotEmpty(accessCode))
+                                    .arg("encryptedAccessCode", encryptedAccessCode)
+                                    .arg("encryptedSalt", encryptedSalt)
+                                    .arg("encryptedJSessionId",encryptedJSessionId);
+
+                        } catch (Exception ex) {
+                            log.error(ex.getMessage());
+                            return getView("denied");
+                        }
+                    } else {
+                            return getView("notfound");
                     }
+                } else {
+                    return getView("notfound");
+                }
+            }
+        };
+    }
+
+    public EasyShareUnrestrictedRunner buildUnrestrictedRunnerStream(final String jSessionId, final String accessCode, final String docId) {
+
+        return new EasyShareUnrestrictedRunner() {
+            @Override
+            public Object run(CoreSession session, IdRef docRef) throws NuxeoException {
+                if (session.exists(docRef)) {
+                    DocumentModel doc = session.getDocument(docRef);
+                    try (OperationContext ctx = new OperationContext(session)) {
+                        DocumentModel docShare = session.getDocument(new IdRef(docId));
+
+                        if (!checkIfShareIsValid(docShare)) {
+                            return Response.serverError().status(Response.Status.NOT_FOUND).build();
+                        }
+
+                        if (!checkIfAccessCodeIsValid(docShare,accessCode)) {
+                            return Response.serverError().status(Response.Status.FORBIDDEN).build();
+                        }
+                        
+                        if (StringUtils.isBlank(jSessionId) || !jSessionId.equals(getRequestJSESSIONID())) {
+                            return Response.serverError().status(Response.Status.FORBIDDEN).build();
+                        }
```

Comment:
Doesn't it a backward compatibility change? What happens if the query parameter for jSessionId is not present?

Reply:
No backward. the case is :

- Retrieve the JSESSIONID when a client "log in" providing the right accessCode and put it as a URL parameter (j), this information will be propagated for all generated links (browse in folders or download content)
- When connected, the user will click on a link to fetch folder content or download a document (the link is dynamically generated on each refresh)
- On click, retrieve the URL parameter j (previously encoded with the original JSESSIONID) and compare it with the current JSESSIONID
- If encoded URL JSESSIONID and current JSESSIONID do not match, we can consider the client (browser) is not the same (maybe the URL has been copied and transferred to another desktop, ...) and so refused the access